include/deal.II/base/vectorization.h

```diff
@@ -3941,6 +3941,167 @@ operator<<(std::ostream &out, const VectorizedArray<Number, width> &p)
   return out;
 }
 
+
+/**
+ * @name Ternary operations on VectorizedArray
+ */
+//@{
+
+/**
+ * TODO
+ */
+enum SIMDComparison : int
+{
+  equal                 = _CMP_EQ_OQ,
+  not_equal             = _CMP_NEQ_OQ,
+  less_than             = _CMP_LT_OQ,
+  less_than_or_equal    = _CMP_LE_OQ,
+  greater_than          = _CMP_GT_OQ,
+  greater_than_or_equal = _CMP_GE_OQ
+};
+
+
+/**
+ * TODO
+ */
+template <SIMDComparison predicate, typename Number>
+DEAL_II_ALWAYS_INLINE inline Number
+compare_and_apply_mask(const Number &left,
+                       const Number &right,
+                       const Number &true_value,
+                       const Number &false_value)
+{
+  bool mask;
+
+  switch (predicate)
+    {
+      case SIMDComparison::equal:
+        mask = left == right;
+      case SIMDComparison::not_equal:
+        mask = left != right;
+      case SIMDComparison::less_than:
+        mask = left < right;
+      case SIMDComparison::less_than_or_equal:
+        mask = left <= right;
+      case SIMDComparison::greater_than:
+        mask = left > right;
+      case SIMDComparison::greater_than_or_equal:
+        mask = left >= right;
+    }
+
+  return mask ? true_value : false_value;
+}
+
+//@}
+
+#ifndef DOXYGEN
+#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__AVX512F__)
+
+template <int predicate>
+DEAL_II_ALWAYS_INLINE inline VectorizedArray<float, 16>
+compare_and_apply_mask(const VectorizedArray<float, 16> &left,
+                       const VectorizedArray<float, 16> &right,
+                       const VectorizedArray<float, 16> &true_values,
+                       const VectorizedArray<float, 16> &false_values)
+{
+  const auto mask = _mm512_cmp_ps(left.data, right.data, predicate);
+  VectorizedArray<float, 16> result;
+  result.data = _mm512_or_ps(_mm512_and_ps(mask, true_values.data),
```

Comment:
Could you in principle write this as
```
  VectorizedArray<...> result {{_mm512_or_ps(...)}};
  return result;
```
possibly with a different number of curly braces? I suspect that the `.data` member of `result` isn't initialized above anyway, so there is no duplicate work -- but this would make it really clear.

Reply:
We currently have no constructor that would support this syntax.

This would be a nice improvement for the class, though :-) @kronbichler *ping*

Reply:
@kronbichler I am not sure about this. We can have private constructors which take `__m512` (for AVX512 and other intrinsics types for other ISAs). The reason why I think it should be private is that users should not be encouraged to use it (since it makes the user code not portable any more - which is actually one of the goals of `VectorizedArray`).

Reply:
@peterrum  You're right - let's not expose implementation details.